### PR TITLE
feat(bungee-hermes): new plugin bungee-hermes

### PR DESCRIPTION
**Pull request by eduv09**
BUNGEE provides a way to create blockchain snapshots and views for different distributed ledgers currently supported by Cacti. The implementation follows the paper "BUNGEE: Dependable Blockchain Views for Interoperability" at https://dl.acm.org/doi/pdf/10.1145/3643689.

To facilitate the integration with multiple different networks, the plugin implements a strategy design pattern. 
It keeps a map <name, Strategy>. Each strategy is capable of capturing asset state proofs, by querying and processing data from the correspondent cactus-ledger-connectors, which is then used by the plugin to create the desired snapshots and views . 
Note that, so far, only strategies for Fabric and Besu networks were implemented.

Any comments or recommendations are appreciated.

### CODE REVIEW

**Comment by RafaelAPB** [comment]
@outSH @petermetz @VRamakrishna dear fellow maintainers I would appreciate a fast review of this PR, which is one of the necessary PRs that refactor the SATP-Hermes implementation. BUNGEE implements the concept of blockchain view (albeit a bit differently than the spec that @VRamakrishna and colleagues created).

File: packages/cactus-plugin-bungee/package.json
Comment (outSH) [inline packages/cactus-plugin-bungee/package.json:3]: Fix verison and provide more detailed description please

File: packages/cactus-plugin-bungee/src/main/json/openapi.json
Comment (outSH) [inline packages/cactus-plugin-bungee/src/main/json/openapi.json:6]: Fix version

**Review by outSH — CHANGES_REQUESTED** [review]

**Review by RafaelAPB — CHANGES_REQUESTED** [review]

File: packages/cactus-plugin-bungee/src/main/typescript/web-services/client-endpoint.ts
Comment (petermetz) [inline packages/cactus-plugin-bungee/src/main/typescript/web-services/client-endpoint.ts:101]: @eduv09 Please refactor the error handling to the new pattern we've (recently) established so that it is more robust and handles HTTP specific error codes automatically:

You can see an example (that is pretty much one to one copy-pastable when it comes to the `catch` block) here in this file on the main branch:

`packages/cactus-core/src/main/typescript/web-services/get-open-api-spec-v1-endpoint-base.ts`

File: packages/cactus-plugin-bungee/src/main/typescript/web-services/client-endpoint.ts
Comment (petermetz) [inline packages/cactus-plugin-bungee/src/main/typescript/web-services/client-endpoint.ts:27]: @eduv09 The name of the endpoint should be based on a verb, e.g. "DoSomethingEndpointV1" instead of "ThingEndpointV1" (where Thing = 'Client' in the specific context).
Please rename it to `CreateViewEndpointV1` (I looked at the code of what it does and that seems to sum it up pretty well).

P.S.: I just saw that you also have an inconsistent naming issue where the endpoint class constructor options are called `CreateViewEndpointOptions`. This is a step in the right direction because there it is a verb, but still, we need to make sure that the options interface and the class name itself are consistent so please make sure to streamline in like this:
`CreateViewEndpointV1Options` and `CreateViewEndpointV1`

File: packages/cactus-plugin-bungee/src/main/typescript/plugin-bungee.ts
Comment (petermetz) [inline packages/cactus-plugin-bungee/src/main/typescript/plugin-bungee.ts:50]: ing.@eduv09 Please be more specific with the name of the plugin (and consequently the package and everything else). I know that the more verbose names are annoying but they are necessary. I wrote a little essay on why this is and will paste it below, please skim through it and then it will hopefully make more sense why we have to insist on more verbose names!

Some (strictly just) suggestions/ideas for a name: 
PluginBungeeSatpHermes
PluginBungeeHermes
PluginBungeeSatp

I realize that this plugin might not be stirctly tied to SATP nor the Hermes implementation, but we just need to give it some name that associates it with either the creators or the other packages that will be using it. It could even be called PluginBungeeRafael as far as I'm concerned because Rafael was the first person to tell me about BUNGEE a few years back. Joking but only half joking. For the sake of namespacing we got to put something there.
My favorite is `PluginBungeeHermes` but you pick whichever you like!

The promised essay on the why:

===

The naming convention for plugins is Plugin__$FUNCTIONALITY_GROUP$____$IMPLEMENTATION_FLAVOR$__

For example PluginKeychainVault where the $FUNCTIONALITY_GROUP$ is `Keychain` and the implementation flavor is `Vault` which allows
to be several different implementation flavors such as the in-memory keychain plugin and all the other ones that we will add
in the future (we also have AWS, GCP and Azure implementations).

Another example is the PluginLedgerConnectorBesu where the $FUNCTIONALITY_GROUP$ is `LedgerConnector` and the flavor
is `Besu`.

This hierarchical naming structure allows us to "scale" the project with several different kinds of implementations for
the same thing.

If you name something as simply PluginSomething (like PluginBUNGEE) then it becomes a squatter on that
namespace in the sense that if anyone wants to implement the same kind of functionality in a different
plugin later and they name theirs "PluginBUNGEEPetersCoolImplementation" then it will get confusing
because people looking at these classnames will rightfully assume that the two are in the same package
or that the original PluginBUNGEE is in some sort of parent relationship to the newer one, but the are
actually peers/siblings/etc and are both in a parent child relationship with the API server that loads
the plugins and the core API package that defines the interfaces between the API server and the plugins.

File: packages/cactus-plugin-bungee/src/main/typescript/public-api.ts
Comment (petermetz) [inline packages/cactus-plugin-bungee/src/main/typescript/public-api.ts:4]: @eduv09 Just marking it here too that after the rename please don't forget to rename the imported files/paths, the classes that are being exported as well (it will be annoying to refactor it everywhere but much easier if you use the search function of VSCode wisely. :-)

File: packages/cactus-plugin-bungee/src/main/typescript/public-api.ts
Comment (petermetz) [inline packages/cactus-plugin-bungee/src/main/typescript/public-api.ts:6]: @eduv09 I'm guessing this factory function got copied over from the besu connector, which is totally cool (I also copy paste older plugins when writing a new one) but please make sure to rename it to match existing patterns. Let me know if you need any more help on this.

The reason why this is important is because the API server imports only one specific name from plugin packages when it is instantiating the plugin, so it is important to comply with this otherwise the API server won't be able to load your plugin implementation at all.

File: packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/package.json
Comment (petermetz) [inline packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/package.json:40]: @eduv09 Please remove the ones from here that aren't strictly necessary for the test case to remain operational (such as chai, mocha, nyc, sinin, ts-node, tslint and most likely typescript as well). The point being that these dependencies will have to be maintained and upgraded because our security automation tooling will catch vulnerabilities in them in the future and then it adds the burden on everybody else later to fix those so it's better to strip down these exmaples that are just for automated tests as much as possible.

File: packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/.gitignore
Comment (petermetz) [inline packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/.gitignore:16]: @eduv09 I'm guessing that this file can be removed because it won't be used at all in the test automation context. Please either remove it or just advise why it is necessary and if there is a legit reason for it to remain then I'm totally OK with that too we just have to have it documented on the purpose of leaving it under version control.

File: packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/tslint.json
Comment (petermetz) [inline packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/tslint.json:23]: @eduv09 I'm guessing that this file can be removed because it won't be used at all in the test automation context. Please either remove it or just advise why it is necessary and if there is a legit reason for it to remain then I'm totally OK with that too we just have to have it documented on the purpose of leaving it under version control.

File: packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/package.json
Comment (petermetz) [inline packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/package.json:18]: @eduv09 I'd also prune the unused scripts here.

File: packages/cactus-plugin-bungee/README.md
Comment (petermetz) [inline packages/cactus-plugin-bungee/README.md:3]: @eduv09 The abstract of the paper mentions what BUNGEE stands for and I'd love it if you included that here as well just so that people get a quicker understanding of it from the full expression.

File: packages/cactus-plugin-bungee/tsconfig.json
Comment (petermetz) [inline packages/cactus-plugin-bungee/tsconfig.json:12]: @eduv09 The current default here is `node` so if you set it to Node16 it might (as in most likely) break the build and some other things. If there is a specific reason why this has to be set to this then let's talk about it and figure out a solution together. Me and @outSH spent countless hours working through issues related to ESM resolutions for example and 99% of the time between the two of us there will be some solution that we can recommend without having to set this to Node16 and potentially making trouble for the build process.

File: packages/cactus-plugin-bungee/tsconfig.json
Comment (petermetz) [inline packages/cactus-plugin-bungee/tsconfig.json:9]: @eduv09 Does this have to be set this way or could it be set to `false`? If it is this way for a specific reason then let's talk about it and try to fix it together os that it doesn't have to be disabled (the lib checking)

**Review by petermetz — CHANGES_REQUESTED** [review]
@eduv09 Huge diff, really difficult to review, sorry in advance for a lot of back and forth. Please see my suggestions, change requests above.
If you need any help just let me know I'm more than happy to get in and contribute to your PR by making some changes that I'm recommending.

**Comment by petermetz** [comment]
@eduv09 I was able to fix the Node16 issue by inheriting from the correct tsconfig.json file in the root:

image


I checked also if we can do lib check but that's not going to work unfortunately so it's OK to leave it turned off ;-(

**Comment by petermetz** [comment]
@eduv09 Also, if you'd like me to help out with some of the code please make sure to invite me to your fork as a collaborator with permissions

File: packages/cactus-plugin-bungee/src/main/typescript/plugin-bungee.ts
Comment (RafaelAPB) [reply packages/cactus-plugin-bungee/src/main/typescript/plugin-bungee.ts:50]: I agree with Peter. PluginBungeeHermes sounds good to me. @eduv09 choose whatever you like the most

File: packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/package.json
Comment (RafaelAPB) [reply packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/package.json:40]: @petermetz is there a way to automatically identify unused packages? [i found depcheck](https://stackoverflow.com/questions/22675725/how-to-find-unused-packages-in-package-json), but does not seem canonical to the project. We could add a dependency, add an npm task to check unused deps, and update the PR for contributors to run depcheck before opening a PR.

File: packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/.gitignore
Comment (RafaelAPB) [reply packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/.gitignore:16]: I believe this chaincode is used to test BUNGEE

Comment (RafaelAPB) [reply packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/.gitignore:16]: Oh, you mean the lint, specifically. Yes, strictly speaking we just need the chaincode.

**Comment by RafaelAPB** [comment]
> also if we can do lib check but that's not going to work unfortunately so it's OK to leave it turned off ;-(
Thank you for your review and service, @petermetz !

**Comment by RafaelAPB** [comment]
I should add another feature that we forgot: the support for creating views on plugin-packages/cactus-plugin-ledger-connector-ethereum. Should be quite similar to Besu. We need this because if we cover evm + fabric + besu we will cover much of the interoperation needs.

File: packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/package.json
Comment (petermetz) [reply packages/cactus-plugin-bungee/src/test/typescript/fabric-contracts/simple-asset/chaincode-typescript/package.json:40]: @RafaelAPB I agree on all of those. Some more context is needed though: We tried depcheck many years ago and it would identify packages to remove that were not actually removable, e.g. we acted on it's advice and the build and/or tests broke.

With all that said, this was close to 5 years ago so the tool could've improved in the meantime and our build process also definitely improved in the meantime (we went through several major upgrades on npm and then yarn as the package manager).

Therefore trying it again seems like a rational thing to do because it could easily be that nowadays "it just works". 

Also especially agree that if we do find a tool (depcheck or any other) that correctly identifies extraneous dependencies then that should be part of the CI so that it's not possible to accidentally commit such dependencies to main anymore.

File: packages/cactus-plugin-bungee/tsconfig.json
Comment (petermetz) [reply packages/cactus-plugin-bungee/tsconfig.json:12]: > @eduv09 The current default here is `node` so if you set it to Node16 it might (as in most likely) break the build and some other things. If there is a specific reason why this has to be set to this then let's talk about it and figure out a solution together. Me and @outSH spent countless hours working through issues related to ESM resolutions for example and 99% of the time between the two of us there will be some solution that we can recommend without having to set this to Node16 and potentially making trouble for the build process.

@eduv09 Please see my comments on the bottom, I've investigated this in the meantime and found a solution.

File: packages/cactus-plugin-bungee/tsconfig.json
Comment (petermetz) [reply packages/cactus-plugin-bungee/tsconfig.json:9]: @eduv09 Also investigated this one and this is fine as-is, resolving.

**Comment by petermetz** [comment]
> > also if we can do lib check but that's not going to work unfortunately so it's OK to leave it turned off ;-(
> 
> Thank you for your review and service, @petermetz !

@RafaelAPB My pleasure! Thank you for helping out with the review as well, this is a big one! :-)

**Comment by eduv09** [comment]
@RafaelAPB @petermetz thankyou for the comments ! I worked on the changes you requested, I hope I accurately changed everything :). 

Some notes:

 - I did not add the Strategy for ethereum connector yet, will work on it soon. Maybe this can/should be done in a separate commit or even pr, since it is a new 'feature'? I will create the strategy regardless.

 - I also did not add the View proof suggested by @RafaelAPB yet, I'll check it as soon as possible.

 - I created one more endpoint "GetAvailableStrategies" besides the one suggested, thought it might be relevant.

 - Lastly, I had a few issues deploying the solidity smart contract in Besu, after changing it to include the ITraceableContract.sol interface. I included the interface file for reference, but the tests are running using the original lockAsset.sol contract (without implementing the interface). 

Thanks! Plugin is now bungee-hermes :100:

File: packages/cactus-plugin-bungee/src/main/typescript/view-creation/proof.ts
Comment (eduv09) [reply packages/cactus-plugin-bungee/src/main/typescript/view-creation/proof.ts]: My idea when creating this Type was to make it general purpose. That's why mspid and signature are optional properties. 
TransactionProof has, at least, a Proof which is the information accessible about the transaction creator (id, mspid, transaction-signature). In the case of Fabric, a TransactionProof will also include a set of Proof, which represent endorsers (id, mspid, endorsement-for-that-transaction).

**Review by petermetz — COMMENTED** [review]
@eduv09 Please dont' forget to update the naming everywhere. GitHub PR title. commit subject, docs, package.json, readme, just to name a few. VSCode free text search is what I usually use for this sort of thing by doing multiple searches for the hypenated versions the camel case versions, etc.

File: packages/cactus-plugin-bungee-hermes/src/main/json/openapi.json
Comment (petermetz) [inline packages/cactus-plugin-bungee-hermes/src/main/json/openapi.json:23]: @eduv09 Please fix the custom check that is failing here:

```sh
ERROR: packages/cactus-plugin-bungee-hermes/src/main/json/openapi.json "info.license.name" must be Apache-2.0
```

**Review by petermetz — CHANGES_REQUESTED** [review]

**Comment by RafaelAPB** [comment]
LGTM. I pushed a rebase to your branch @eduv09 . Please work on adding ethereum connector support / improvements over proofs in a following PR. Let's wait for @petermetz's comments before merging.

**Comment by RafaelAPB** [comment]
Please also add me as a maintainer of this package, @eduv09 , by changing the file Codeowners: https://github.com/hyperledger/cacti/blob/main/CODEOWNERS

File: packages/cactus-plugin-bungee-hermes/src/main/typescript/view-creation/proof.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-bungee-hermes/src/main/typescript/view-creation/proof.ts:33]: Can this be removed, if not required?

File: packages/cactus-plugin-bungee-hermes/src/main/typescript/view-creation/proof.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-bungee-hermes/src/main/typescript/view-creation/proof.ts:37]: Can we add a quick in-code comment here describing the nature of creator for the ledgers? It would be very helpful I believe, something like

The term creator refers to `endorsementID` for fabric ledger or `txTx.data.transaction.from` in case of besu ledger

File: packages/cactus-plugin-bungee-hermes/src/main/typescript/view-creation/transaction-proof.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-bungee-hermes/src/main/typescript/view-creation/transaction-proof.ts:4]: Can we call `creator` class var as something different (Like how you have used the term `blockCreator`)? Because Proof itself has a creator class var

File: packages/cactus-plugin-bungee-hermes/src/test/typescript/integration/fabric-test-pruning.test.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-bungee-hermes/src/test/typescript/integration/fabric-test-pruning.test.ts:173]: Can we remove the commented code, if not required?

**Review by jagpreetsinghsasan — COMMENTED** [review]
Looks good overall, specially how generic the plugin class parameters are designed to support multiple ledgers.

**Comment by eduv09** [comment]
@jagpreetsinghsasan all done I think. Thanks for the comments !
@petermetz I hope I changed the naming everywhere this time. If you detect it somewhere else that is worth changing, let me know. 
For refference:
 - Last push (4e5...) was a rebase against new main.
 - [02a35c8] Changes requested from @jagpreetsinghsasan 
 - [061818a] Changes requested from @petermetz

**Comment by RafaelAPB** [comment]
> @jagpreetsinghsasan all done I think. Thanks for the comments ! @petermetz I hope I changed the naming everywhere this time. If you detect it somewhere else that is worth changing, let me know. For refference:
> 
> * Last push (4e5...) was a rebase against new main.
> * [[02a35c8](https://github.com/hyperledger/cacti/commit/02a35c8630a2abf374974496340aa9c59462d4b0)] Changes requested from @jagpreetsinghsasan
> * [[061818a](https://github.com/hyperledger/cacti/commit/061818a34f913ed48c0cdad227a925d00a4b4b2a)] Changes requested from @petermetz

Great work, @eduv09. Please set the addressed review comments as "resolved".
All LGTM for merge.

**Review by RafaelAPB — APPROVED** [review]

File: packages/cactus-plugin-bungee-hermes/src/main/typescript/view-creation/proof.ts
Comment (RafaelAPB) [reply packages/cactus-plugin-bungee-hermes/src/main/typescript/view-creation/proof.ts:42]: +1

**Comment by jagpreetsinghsasan** [comment]
LGTM

**Comment by RafaelAPB** [comment]
@eduv09 please mark suggestions as solved and rebase so we can merge. Good work!

**Comment by eduv09** [comment]
Done @RafaelAPB

File: packages/cactus-plugin-bungee/README.md
Comment (petermetz) [reply packages/cactus-plugin-bungee/README.md]: @eduv09 I see that you resolved this. Could you point out where in the README the acronym's full meaning is shown?

**Review by petermetz — CHANGES_REQUESTED** [review]

**Comment by RafaelAPB** [comment]
@eduv09 please fix the conflicts that appear.

File: packages/cactus-plugin-bungee/README.md
Comment (eduv09) [reply packages/cactus-plugin-bungee/README.md]: image
image


The changes were introduces in commit e319e6c, if I am not mistaken.

**Review by petermetz — APPROVED** [review]
LGTM

**Comment by eduv09** [comment]
@petermetz there were two old generated files to be removed. I removed them now, It should fix yarn_codegen . Sorry for that

**Comment by AndreAugusto11** [comment]
closes #3106

**Pull request merged**
